pkg/registry/core/service/storage/rest.go

```diff
@@ -574,8 +572,14 @@ func (r *REST) getAllocatorBySpec(service *api.Service) ipallocator.Interface {
 	}
 
 	secondaryAllocatorCIDR := r.secondaryServiceIPs.CIDR()
-	if *(service.Spec.IPFamily) == api.IPv6Protocol && netutil.IsIPv6CIDR(&secondaryAllocatorCIDR) {
-		return r.secondaryServiceIPs
+	if netutil.IsIPv6CIDR(&secondaryAllocatorCIDR) {
+		if *(service.Spec.IPFamily) == api.IPv6Protocol {
+			return r.secondaryServiceIPs
+		}
+	} else {
```

Comment:
The absence of this `else` meant that Dual Stack IPv6 primary clusters would allocate an IPv6 address when family was IPv4 (because the primary serviceIP allocator was IPv6)